src/main/scala/com/fulcrumgenomics/commons/util/ConfigurationLike.scala

```diff
@@ -0,0 +1,247 @@
+
+import java.nio.file.Path
+import java.time.Duration
+
+import com.fulcrumgenomics.commons.CommonsDef._
+import com.fulcrumgenomics.commons.io.PathUtil.pathTo
+import com.fulcrumgenomics.commons.reflect.ReflectionUtil
+import com.typesafe.config.{Config, ConfigFactory, ConfigParseOptions}
+
+import scala.collection.SortedSet
+import scala.reflect.runtime.universe.{TypeTag, typeOf}
+import scala.util.{Failure, Success}
+
+/**
+  * Trait that provides useful methods for resolving all kinds of things in configuration.  Uses
+  * [[Config]] to retrieve configuration values in a type-safe way.
+  *
+  * Keeps track of all configuration keys that have been requested so that they can be reported later if desired.
+  *
+  * @example Retrieve a value at a given path, where the path must exist:
+  * {{{
+  *   scala> import com.fulcrumgenomics.commons.util.ConfigurationLike
+  *   scala> import com.typesafe.config.{Config, ConfigFactory}
+  *   scala> val conf = new ConfigurationLike {
+  *     val config: Config = ConfigFactory.parseString("""a.str = "string", a.int = 2, a.set = [1,2,3]""")
+  *   }
+  *   scala> conf[String]("a.str")
+  *   res0: String = string
+  *   scala> conf[Int]("a.int")
+  *   res1: Int = 2
+  *   scala> conf[Set[Int]]("a.set")
+  *   res2: Set[Int] = Set(1, 2, 3)
+  *   scala> conf[Set[Int]]("a.does-not-exist")
+  *   java.lang.IllegalArgumentException: Exception retrieving configuration key 'a.does-not-exist': No configuration setting found for key 'a.does-not-exist'
+  *   ...
+  * }}}
+  *
+  * @example Retrieve a value at a given path, where the path may not exist
+  * {{{
+  *   scala> import com.fulcrumgenomics.commons.util.ConfigurationLike
+  *   scala> import com.typesafe.config.{Config, ConfigFactory}
+  *   scala> val conf = new ConfigurationLike {
+  *     val config: Config = ConfigFactory.parseString("""a.str = "string", a.int = 2, a.set = [1,2,3]""")
+  *   }
+  *   scala> conf.get[String]("a.str")
+  *   res0: Option[String] = Some(string)
+  *   scala> conf.get[Int]("a.int")
+  *   res1: Option[Int] = Some(2)
+  *   scala> conf.get[Set[Int]]("a.set")
+  *   res2: Option[Set[Int]] = Some(Set(1, 2, 3))
+  *   scala> conf.get[Set[Int]]("a.does-not-exist")
+  *   res3: Option[Set[Int]] = None
+  * }}}
+  *
+  * @example Retrieve a value at a given path if it exists, otherwise return a default:
+  * {{{
+  *   scala> import com.fulcrumgenomics.commons.util.ConfigurationLike
+  *   scala> import com.typesafe.config.{Config, ConfigFactory}
+  *   scala> val conf = new ConfigurationLike {
+  *     val config: Config = ConfigFactory.parseString("""a.str = "string", a.int = 2, a.set = [1,2,3]""")
+  *   }
+  *   scala> conf.getOrElse[String]("a.does-not-exist", "default")
+  *   res0: String = default
+  *   scala> conf.getOrElse[Int]("a.does-not-exist", 0)
+  *   res1: Int = 0
+  *   scala> conf.getOrElse[Set[Int]]("a.does-not-exist", Set.empty)
+  *   res2: Set[Int] = Set()
+  * }}}
+  */
+trait ConfigurationLike {
```

Comment:
This is lifted over from dagr, but with quite a few changes.  I'm really not sure about the move from `def configure()` to `apply()/get()`.  The intent of the trait in dagr is that you could mix it into just about any class to gain access to configuration and there it felt natural to write things like:

```scala
private val limit = configure[Int]("some_limit")
```

Changing it to use `apply()` means that any class that wants to mix in the trait suddenly has a public apply() method that accesses configuration.  Which makes it hard to the class to provide it's own apply() method(s) that suit it's own purposes and also makes for a confusing public API. 

It feels like the intent here is very different - to make it easier to generate `Config` singleton objects?  In which case maybe it should not be a trait but instead be an single class that can be instantiated (e.g. into a case object) or extended to add more functionality?